### PR TITLE
Victor/refactor/az span

### PR DESCRIPTION
**Pull request by vhvb1989**
https://github.com/Azure/azure-sdk-for-c/issues/243

### CODE REVIEW

File: sdk/core/core/inc/az_pair.h [hunk 1]
```diff
@@ -69,7 +69,7 @@ az_pair_span_as_writer(az_pair_span const * const self, az_pair_action const wri
 AZ_ACTION_FUNC(az_pair_span_as_writer, az_pair_span const, az_pair_writer)
 
 AZ_NODISCARD AZ_INLINE bool az_pair_span_eq(az_pair const a, az_pair const b) {
```
Comment (JeffreyRichter) [inline sdk/core/core/inc/az_pair.h]: s/b az_pair_span_is_equal to match az_span_is_equal

File: sdk/core/core/src/az_json_builder.c [hunk 2]
```diff
@@ -200,9 +200,7 @@ az_json_builder_write(az_json_builder * const self, az_json_token const value) {
       self->need_comma = true;
       return az_json_builder_write_span(self, value.data.span);
     }
-    default: {
-      return AZ_ERROR_ARG;
-    }
+    default: { return AZ_ERROR_ARG; }
```
Comment (JeffreyRichter) [inline sdk/core/core/src/az_json_builder.c]: Is this a contract violation that can return error, or break into the debugger, or be ignored?
If so, this code doesn't accomplish all of this.

File: sdk/core/core/src/az_span.c [hunk 3]
```diff
@@ -18,7 +18,7 @@ AZ_NODISCARD AZ_INLINE az_result_byte az_ascii_lower(az_result_byte const value)
   return 'A' <= value && value <= 'Z' ? value + AZ_ASCII_LOWER_DIF : value;
 }
 
-AZ_NODISCARD bool az_span_eq_ascii_ignore_case(az_span const a, az_span const b) {
+AZ_NODISCARD bool az_span_is_ascii_ignore_case(az_span const a, az_span const b) {
```
Comment (JeffreyRichter) [inline sdk/core/core/src/az_span.c]: This name still isn't good enough. How about az_span_is_equal_ignoring_case

File: sdk/core/core/src/az_json_builder.c (hunk 2 above)
Comment (vhvb1989) [reply sdk/core/core/src/az_json_builder.c]: @sergey-shandar  would have an answer here

Comment (vhvb1989) [reply sdk/core/core/src/az_json_builder.c]: https://github.com/Azure/azure-sdk-for-c/issues/290

File: sdk/core/core/inc/az_span.h [hunk 4]
```diff
@@ -102,15 +102,13 @@ AZ_NODISCARD AZ_INLINE az_result az_error_unexpected_char(az_result_byte const c
...
-#define AZ_ARRAY_SIZE(ARRAY) (sizeof(ARRAY) / sizeof(*ARRAY))
+AZ_NODISCARD az_result az_span_to_uint64(az_span const self, uint64_t * const out);
 
 #define AZ_SPAN_FROM_ARRAY(ARRAY) \
-  { .begin = ARRAY, .size = AZ_ARRAY_SIZE(ARRAY) }
+  { .begin = ARRAY, .size = (sizeof(ARRAY) / sizeof(*ARRAY)) }
 
-AZ_NODISCARD AZ_INLINE az_span az_span_from_one(uint8_t const * ptr) {
-  return (az_span){ .begin = ptr, .size = 1 };
+AZ_NODISCARD AZ_INLINE az_span az_span_from_buffer_range(uint8_t const * ptr, size_t const size) {
```
Comment (sergey-shandar) [inline sdk/core/core/inc/az_span.h]: `az_span_from_one` should be used when we have a `uint8_t` variable. In this case, `size` doesn't make sense. If you need a function for sub span, use `az_span_drop`, `az_span_take`, `az_span_sub`.

**Pull request merged**